### PR TITLE
Create secret for basic login

### PR DESCRIPTION
**Pull request by gabrielwen**
Relates to #2661

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/2690)

### CODE REVIEW

**Comment by gabrielwen** [comment]
PR is ready.

File: bootstrap/pkg/client/gcp/gcp.go [hunk 1]
```diff
@@ -719,7 +723,7 @@ func (gcp *Gcp) writeGcpSecret(client *clientset.Clientset, secretName string, d
 	return err
 }
 
-func (gcp *Gcp) createGcpSecret(ctx context.Context, client *clientset.Clientset,
+func (gcp *Gcp) createGcpServiceAcctSecret(ctx context.Context, client *clientset.Clientset,
```
Comment (lluunn) [inline bootstrap/pkg/client/gcp/gcp.go]: please add comment for func (and below).
Like what's created

File: bootstrap/cmd/kfctl/cmd/apply.go [hunk 2]
```diff
@@ -43,8 +43,10 @@ var applyCmd = &cobra.Command{
 			return
 		}
 		options := map[string]interface{}{
-			string(kftypes.OAUTH_ID):     applyCfg.GetString(string(kftypes.OAUTH_ID)),
-			string(kftypes.OAUTH_SECRET): applyCfg.GetString(string(kftypes.OAUTH_SECRET)),
+			string(kftypes.OAUTH_ID):            applyCfg.GetString(string(kftypes.OAUTH_ID)),
```
Comment (lluunn) [inline bootstrap/cmd/kfctl/cmd/apply.go]: It's fine for now, but this should probably not be passed at apply.

File: bootstrap/pkg/client/gcp/gcp.go [hunk 3]
```diff
@@ -802,13 +790,87 @@ func (gcp *Gcp) createSecrets(options map[string]interface{}) error {
...
+	userEmail := getSA(gcp.GcpApp.Name, "user", gcp.GcpApp.Spec.Project)
+	if err := gcp.createGcpServiceAcctSecret(ctx, k8sClient, adminEmail, ADMIN_SECRET_NAME); err != nil {
+		return fmt.Errorf("cannot create admin secret %v Error %v", ADMIN_SECRET_NAME, err)
+
+	}
+	if err := gcp.createGcpServiceAcctSecret(ctx, k8sClient, userEmail, USER_SECRET_NAME); err != nil {
+		return fmt.Errorf("cannot create user secret %v Error %v", USER_SECRET_NAME, err)
+
+	}
+	if err := gcp.createIapSecret(ctx, k8sClient, options); err != nil {
```
Comment (lluunn) [inline bootstrap/pkg/client/gcp/gcp.go]: shouldn't this be done conditionally (!UseBasicAuth?)
same as below

File: bootstrap/pkg/client/gcp/gcp.go [hunk 4]
```diff
@@ -802,13 +790,87 @@ func (gcp *Gcp) createSecrets(options map[string]interface{}) error {
...
+	}
+	if username == "" {
+		return fmt.Errorf("At least one of --%v or ENV `%v` needs to be set.",
+			string(kftypes.BASIC_AUTH_USERNAME), BASIC_AUTH_USERNAME)
+	}
+
+	password := ""
+	if options[string(kftypes.BASIC_AUTH_PASSWORD)] != nil &&
+		options[string(kftypes.BASIC_AUTH_PASSWORD)].(string) != "" {
+		password = options[string(kftypes.BASIC_AUTH_PASSWORD)].(string)
```
Comment (lluunn) [inline bootstrap/pkg/client/gcp/gcp.go]: why do we have kftypes.BASIC_AUTH_PASSWORD and BASIC_AUTH_PASSWORD?

File: bootstrap/pkg/client/gcp/gcp.go (hunk 1 above)
Comment (gabrielwen) [reply bootstrap/pkg/client/gcp/gcp.go]: added.

File: bootstrap/pkg/client/gcp/gcp.go (hunk 3 above)
Comment (gabrielwen) [reply bootstrap/pkg/client/gcp/gcp.go]: metric-controller also uses it.  added a comment as TODO such that we should check not only `UseBasicAuth`, but also if `metric-controller` is part of components.

File: bootstrap/pkg/client/gcp/gcp.go (hunk 4 above)
Comment (gabrielwen) [reply bootstrap/pkg/client/gcp/gcp.go]: it's part of `CliOptions` (string).

File: bootstrap/pkg/client/gcp/gcp.go (hunk 3 above)
Comment (gabrielwen) [reply bootstrap/pkg/client/gcp/gcp.go]: also, `UseBasicAuth` checks is in `createBasicAuthSecret`

File: bootstrap/pkg/client/gcp/gcp.go (hunk 1 above)
Comment (lluunn) [reply bootstrap/pkg/client/gcp/gcp.go]: write to GCP --> insert to cluster?

File: bootstrap/pkg/client/gcp/gcp.go (hunk 4 above)
Comment (lluunn) [reply bootstrap/pkg/client/gcp/gcp.go]: I mean, why do we duplicate things at two places?

File: bootstrap/pkg/client/gcp/gcp.go (hunk 1 above)
Comment (gabrielwen) [reply bootstrap/pkg/client/gcp/gcp.go]: done.

File: bootstrap/pkg/client/gcp/gcp.go (hunk 4 above)
Comment (gabrielwen) [reply bootstrap/pkg/client/gcp/gcp.go]: was going to use ENV as well, thought twice, removed.

File: bootstrap/cmd/kfctl/cmd/apply.go (hunk 2 above)
Comment (gabrielwen) [reply bootstrap/cmd/kfctl/cmd/apply.go]: thought twice, make them init params and write to app.yaml and encrypt password.

File: bootstrap/cmd/kfctl/cmd/init.go [hunk 5]
```diff
@@ -48,7 +49,20 @@ or a <name>. If just <name> a directory <name> will be created in the current di
 		repo := initCfg.GetString(string(kftypes.REPO))
 		project := initCfg.GetString(string(kftypes.PROJECT))
 		init_gcp := initCfg.GetBool(string(kftypes.SKIP_INIT_GCP_PROJECT))
-		basic_auth := initCfg.GetBool(string(kftypes.USE_BASIC_AUTH))
+
+		useBasicAuth := initCfg.GetBool(string(kftypes.USE_BASIC_AUTH))
```
Comment (lluunn) [inline bootstrap/cmd/kfctl/cmd/init.go]: I think these should not be at init, but at generate.
https://github.com/kubeflow/kubeflow/issues/2691#issuecomment-472970551

Comment (gabrielwen) [reply bootstrap/cmd/kfctl/cmd/init.go]: Let's make a follow-up PR for it.  Want to submit it and test with another PR.

File: bootstrap/pkg/client/gcp/gcp.go (hunk 3 above)
Comment (kunmingg) [reply bootstrap/pkg/client/gcp/gcp.go]: Should be safe to remove IapSecret when UseBasicAuth.
I'll change metric-controller to skip IapSecret when not available.

Comment (gabrielwen) [reply bootstrap/pkg/client/gcp/gcp.go]: Done.

**Pull request merged**